lib/Frontend/Frontend.cpp

```diff
@@ -497,58 +565,131 @@ void CompilerInstance::performSema() {
   if (options.DebugTimeExpressionTypeChecking) {
     TypeCheckOptions |= TypeCheckingFlags::DebugTimeExpressions;
   }
+  return TypeCheckOptions;
+}
+
+// Return true if had load error
+bool CompilerInstance::parsePartialModulesAndLibraryFiles(
+    SourceFile::ImplicitModuleImportKind implicitModuleImportKind,
+    ModuleDecl *objCModuleUnderlyingMixedFramework,
+    ModuleDecl *importedHeaderModule,
+    SmallVectorImpl<ModuleDecl *> &importModules,
+    PersistentParserState &PersistentState,
+    DelayedParsingCallbacks *DelayedParseCB) {
+  SharedTimer timer("performSema-parsePartialModulesAndLibraryFiles");
+  bool hadLoadError = false;
+  // Parse all the partial modules first.
+  for (auto &PM : PartialModules) {
+    assert(PM.ModuleBuffer);
+    if (!SML->loadAST(*MainModule, SourceLoc(), std::move(PM.ModuleBuffer),
+                      std::move(PM.ModuleDocBuffer)))
+      hadLoadError = true;
+  }
+
+  // Then parse all the library files.
+  for (auto BufferID : BufferIDs) {
+    if (BufferID != MainBufferID)
+      parseALibraryFile(BufferID, implicitModuleImportKind,
+                        objCModuleUnderlyingMixedFramework,
+                        importedHeaderModule, importModules, PersistentState,
+                        DelayedParseCB);
+  }
+  if (Invocation.isCodeCompletion()) {
+    // When we are doing code completion, make sure to emit at least one
+    // diagnostic, so that ASTContext is marked as erroneous.  In this case
+    // various parts of the compiler (for example, AST verifier) have less
+    // strict assumptions about the AST.
+    Diagnostics.diagnose(SourceLoc(), diag::error_doing_code_completion);
+  }
+  return hadLoadError;
+}
+
+void CompilerInstance::parseMainAndTypeCheckTopLevelFiles(
```

Comment:
This "and" makes me uncomfortable. Parsing and type-checking are inextricably tied together for a main source file (because we type-check it in sections), but they aren't for other files.

Reply:
OK. (thinking out loud here...) What unites the code in this routine is that it all uses the result of computeTypeCheckingOptions(). This routine could be three: one called 'parseAndTypeCheckTheMainFile', one called 'typeCheckTopLevelInputsExcludingMain', and one called 'typeCheckMainModule'. But what I like about the composition into one is that it unites all the type-checking in performSema into a single function. I could rename it something like 'checkTypesWhileParsingMain'. I'll do that for now. What would be best?

Reply:
Hm, I see the issue. I think I would just want to call it 'performTypeChecking' or something, though, and just mention the parsing of the main file in the comments. (Except not that particular name because that's also the top-level, per-file operation.) Not all targets *have* a main file.